### DISCUSSION TITLE
Does sqlmodel support ORM events as sqlalchemy does?

### DISCUSSION THREAD

**Discussion by khaliullin** [fastapi/sqlmodel#648] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import List, Optional
from sqlalchemy import Column, JSON, event
from sqlmodel import SQLModel, Field, Relationship


class Base(SQLModel):
    id: int = Field(default=None, nullable=False, primary_key=True)

    class Config:
        arbitrary_types_allowed = True


class Task(Base, table=True):
    title: str
    type: str = Field(nullable=False)


# SQLAlchemy ORM Events listener
@event.listens_for(Task.__table__, 'after_commit')  # ideally this should be common table to provide logging action
def receive_after_commit(session):
    # write changes to AuditLog table
```

### Description

I am trying to create audit table for my application. I need to log all changes (at least CREATE, UPDATE actions) to AuditLog database table, to keep changes history of every record with user, action and fields information. Could not find any examples in SQLModel documentation, but SQLAlchemy provides [ORM Events](https://docs.sqlalchemy.org/en/14/orm/events.html) to implement this. Is it possible to do with SQLModel tables?

```
@event.listens_for(Task.__table__, 'after_commit')
def receive_after_commit(session):
    # write changes to AuditLog table
```
this code raises error:
`sqlalchemy.exc.InvalidRequestError: No such event 'after_commit' for target '<class 'app.models.Task'>'`

I need to get sqlalchemy class for sqlmodel Task class. Is it possible to do? If yes, how to get such class for common model to  avoid code duplication?

### Operating System

macOS

### SQLModel Version

0.0.8

### Python Version

3.11.4

**Comment by antont** [comment]
I'd guess it's possible, and that the answer is there, but I could not grasp the metamodel programming enough to get it. https://github.com/tiangolo/sqlmodel/blob/main/sqlmodel/main.py#L209

**Comment by jlantz** [comment, marked as answer]
I'm not sure if something changed or perhaps I'm just targeting a different event at the model level vs the table level, but this code is working for me to trigger events after_insert:

```
from sqlalchemy import event
from .models.job import Job

@event.listens_for(Job, "after_insert")
def handle_job_created(
    mapper,
    connection,
    target,
) -> Job:
    job = target
    ...
```

I previously tried to post code that I thought was working but in fact wasn't. I got tripped up on the jump from my async endpoints to synchronous listener function.

**Reply by vimeh** [reply]
Same here; at the very least can confirm that `listens_for()` on `after_insert` events with the same structure as above worked for me at this time.
